src/cpp/session/modules/SessionDependencies.cpp

```diff
@@ -310,106 +311,97 @@ Error installDependencies(const json::JsonRpcRequest& request,
+         // that the package directory is locked during installation; since this will run in the
+         // background we want reduce the odds of corruption via a competing package install attempt
+         // in another process.
+         script += "utils::install.packages('" + dep.name + "', " +
+                "repos = '"+ module_context::CRANReposURL() + "'";
+
+         if (dep.source) 
+         {
+            // Install from source if requested
+            script += ", type = 'source'";
+         }
+
+         script += ", lock = TRUE)";
       }
       else if (dep.location == kEmbeddedPackageDependency)
       {
          EmbeddedPackage pkg = embeddedPackageInfo(dep.name);
-         if (!pkg.empty())
-            embeddedPackages.push_back(pkg.archivePath);
+
+         // Build install command for bundled archive
+         script += "utils::install.packages('" + pkg.archivePath + 
+            "', repos = NULL, type = 'source', lock = TRUE)";
       }
-   }
 
-   // build install command
-   std::string cmd("{ " + module_context::CRANDownloadOptions() + "; ");
-   if (!cranPackages.empty())
-   {
-      std::string pkgList = boost::algorithm::join(cranPackages, ",");
-      cmd += "utils::install.packages(c(" + pkgList + "), " +
-             "repos = '"+ module_context::CRANReposURL() + "'";
-      cmd += ");";
+      script += "\n\n";
    }
-   if (!cranSourcePackages.empty())
-   {
-      std::string pkgList = boost::algorithm::join(cranSourcePackages, ",");
-      cmd += "utils::install.packages(c(" + pkgList + "), " +
-             "repos = '"+ module_context::CRANReposURL() + "', ";
-      cmd += "type = 'source');";
-   }
-   for (const std::string& pkg : embeddedPackages)
-   {
-      cmd += "utils::install.packages('" + pkg + "', "
-                                      "repos = NULL, type = 'source');";
-   }
-   cmd += "}";
 
-   // build args
-   std::vector<std::string> args;
-   args.push_back("--slave");
+   jobs::ScriptLaunchSpec installJob(
+         // Job name: just the package name if we have one 
+         deps.size() == 1 ? 
+            ("Install '" + deps[0].name + "'") :
+            ("Install R packages"),
 
-   // for packrat projects we execute the profile and set the working
-   // directory to the project directory; for other contexts we just
-   // propagate the R_LIBS
-   if (module_context::packratContext().modeOn)
-   {
-      options.workingDir = projects::projectContext().directory();
-   }
-   else
-   {
-      args.push_back("--vanilla");
-      core::system::Options childEnv;
-      core::system::environment(&childEnv);
-      std::string libPaths = module_context::libPathsString();
-      if (!libPaths.empty())
-         core::system::setenv(&childEnv, "R_LIBS", libPaths);
-      options.environment = childEnv;
-   }
+         // Script to run for job
+         script,
+
+         // Directory in which to run job
+         module_context::packratContext().modeOn ?
```

Comment:
Note: may also want this for `renv` (but I haven't implemented the similar `renvContext()` yet)

Reply:
Thanks, noted!